### ISSUE TITLE
[AzureTools][Aspire] No "Hello World" log shows in the "Console log" window for the Console App that was added later when CTRL+F5ing the project

### ISSUE THREAD

**Issue by v-mengwe** [dotnet/aspire#4489]
**Regression:** This issue **also reproduces** on Aspire 8.0.0

**Clean machine:** Win11 x64 23h2 ENU
**VS Version:** 17.10.3 Preview 1.0 [35012.161.d17.10]
**Aspire Version:** 8.1.0-preview.1.24312.2 
Apply NuGet Feeds
**Podman Desktop:** 1.10.3
**Podman:** 5.1.1

**REPRO STEPS:**

1. Create a new ASP.NET Core MVC project with 'Enlist in .NET Aspire orchestration' checked.
2. Right click on the solution -> Add -> New Project -> Console App.
3. Right click on the Console App project -> Add -> .NET Aspire Orchestration Support -> Click 'OK' when popping the dialog.
    ![image](https://github.com/dotnet/aspire/assets/62053665/eb067acd-f20a-469d-a073-0a281c1e4437)
4. CTRL+F5 -> Click ‘View' under "Logs" of Console App project.
5. Check whether the "Hello World" log shows in the "Console logs" window.

**Expect:**
The "Hello World" log shows in the "Console log" window.
![image](https://github.com/dotnet/aspire/assets/62053665/75c42a7b-1721-4ac7-b83b-b2774449ca76)

**Actual:**
No "Hello World" log shows in the "Console log" window.
![image](https://github.com/dotnet/aspire/assets/62053665/8c0207f4-e616-41c9-ab75-0fece90b00e0)

**More Info:**
1. This issue **doesn't reproduce** when creating Console App in the step 1.
2. This issue **doesn't reproduce** when F5ing the project.
3. This issue **doesn't reproduce** when executing "dotnet run/dotnet watch" for the project.
4. This issue **also reproduces** when changing the ASP.NET Core MVC project to Aspire Starter App in the step 1.
5. On Aspire 8.0.0: This issue **doesn't reproduce** when F5ing the project and then CTRL+F5ing the project.
    On Aspire 8.1.0-preview.1.24312.2: This issue **still reproduces** when F5ing the project and then CTRL+F5ing the project.

**Comment by davidfowl** [comment]
cc @danegsta @karolz-ms

**Comment by karolz-ms** [comment]
Good catch @v-mengwe I was able to repro and it looks like a VS tooling bug to me. 

What I observed:

1. The issue only repros on first try. Subsequent Ctrl-F5 runs work fine.
2. The expected state is that the console app shows in the dashboard as "finished" and "Hello, world!" appears in the logs.
3. When the issue reproduces, the console app shows as "running" in the dashboard and there are no logs available.
4. I see no errors from the DCP-IDE communication in DCP logs.
5. I do see an "unexpected error occurred in Visual Studio" yellow warning in the IDE. This does not happen during successful execution case. When I request more details, the following is shown:

```shell
=====================
6/14/2024 11:13:35 AM
Recoverable
System.AggregateException: One or more errors occurred.
---> (Inner Exception #0) System.InvalidOperationException: Semaphore is already held and reentrancy setting is 'NotAllowed'.
   at Microsoft.Verify.FailOperation(String message, Object[] args)
   at Microsoft.VisualStudio.Threading.ReentrantSemaphore.NotAllowedSemaphore.<ExecuteAsync>d__3`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.VS.HotReload.ProjectHotReloadSessionManager.<OnProcessExitedAsync>d__27.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.CommonProjectSystemTools.Rethrow(Exception ex)
   at Microsoft.VisualStudio.ProjectSystem.ExceptionFilter.<>c__DisplayClass2_0.<Guard>g__Action|0()
   at GuardMethodClass.GuardMethod(Func`1 , Func`2 , Func`2 )
<--- (Inner Exception #0) 
===================
```

Looks like under certain circumstances the IDE is not able to start the console app as requested.

**Comment by BillHiebert** [comment]
@tmeschter

**Comment by tmeschter** [comment]
So the "unexpected error occurred in Visual Studio" would have been fixed by dotnet/project-system#9473. It's likely that is in fact the fix for the underlying issue, but it is also possible that the exception message is incidental and there's some other problem going on.

**Comment by v-mengwe** [comment]
Verified on VS 17.11 preview 3 main-35017.235 with Aspire 8.1.0-preview.1.24307.2  installed, this issue doesn't reproduce on VS 17.11 so close it.
